theatre/studio/src/utils/derive-utils.tsx

```diff
@@ -0,0 +1,109 @@
+import {isDerivation, prism, val} from '@theatre/dataverse'
+import type {IDerivation, Pointer} from '@theatre/dataverse'
+import {useDerivation} from '@theatre/react'
+import type {$IntentionalAny} from '@theatre/shared/utils/types'
+import React, {useMemo, useRef} from 'react'
+import {invariant} from './invariant'
+
+type DeriveAll<T> = IDerivation<
+  {
+    [P in keyof T]: T[P] extends $<infer R> ? R : never
+  }
+>
+
+export type $<T> = IDerivation<T> | Pointer<T>
+
+function deriveAllD<T extends [...$<any>[]]>(obj: T): DeriveAll<T>
+function deriveAllD<T extends Record<string, $<any>>>(obj: T): DeriveAll<T>
+function deriveAllD<T extends Record<string, $<any>> | $<any>[]>(
+  obj: T,
+): DeriveAll<T> {
+  return prism(() => {
+    if (Array.isArray(obj)) {
+      const values = new Array(obj.length)
+      for (let i = 0; i < obj.length; i++) {
+        values[i] = obj[i].getValue()
+      }
+      return values
+    } else {
+      const values: $IntentionalAny = {}
+      for (const k in obj) {
+        values[k] = val((obj as Record<string, $<any>>)[k])
+      }
+      return values
+    }
+  }) as $IntentionalAny
+}
+
+export function useReactPrism(
+  fn: () => React.ReactNode,
+  deps?: any[],
+): React.ReactElement {
+  const derivation = useMemo(() => prism(fn), deps ?? [])
+  return <DeriveElement der={derivation} />
+}
+
+export function reactPrism(fn: () => React.ReactNode): React.ReactElement {
+  return <DeriveElement der={prism(fn)} />
+}
+
+function DeriveElement(props: {der: IDerivation<React.ReactNode>}) {
+  const node = useDerivation(props.der)
+  return <>{node}</>
+}
+
+class TSErrors<M> extends Error {}
```

Comment:
Why define this type? Isn't it the same as `Error`? Needs a doc?

Reply:
Applied comments and made this an `interface` so it's not included